### PR TITLE
Add support for C-style comments

### PR DESCRIPTION
**Pull request by altmannmarcelo**
This commit adds support for C-style comments supported by MySQL. It parses and consumes the optional version number after the `!` character.

### CODE REVIEW

File: src/dialect/mod.rs [hunk 1]
```diff
@@ -898,6 +898,11 @@ pub trait Dialect: Debug + Any {
         false
     }
 
+    /// Returns true if the dialect supports hint and C-style comments
```
Comment (iffyio) [inline src/dialect/mod.rs]: Can we add an example here? I think it won't be clear otherwise what this means by only reading the description

File: src/tokenizer.rs [hunk 2]
```diff
@@ -4070,4 +4101,39 @@ mod tests {
             panic!("Tokenizer should have failed on {sql}, but it succeeded with {tokens:?}");
         }
     }
+    #[test]
+    fn tokenize_multiline_comment_with_c_style_comment() {
+        let sql = String::from("0/*! word */1");
```
Comment (iffyio) [inline src/tokenizer.rs:4280]: Looking at their docs, I'm wondering if/how we support these examples?
```sql
SELECT /*! STRAIGHT_JOIN */ col1 FROM table1,table2
/*!50110 KEY_BLOCK_SIZE=1024 */
SELECT /*! BKA(t1) */ FROM T
```

Comment (altmannmarcelo) [reply src/tokenizer.rs:4280]: @iffyio - Parsing the c_style comment unblocks sqlparser to not discard those as if they were a normal comment. Support for each hint will have to be added in a case by case bases. For example #2033 - MySQL adds a c-style comment if you run SHOW CREATE TABLE:

```
mysql> SHOW CREATE TABLE b;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                    |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+
| b     | CREATE TABLE `b` (
  `ID` int DEFAULT NULL,
  `b` char(1) DEFAULT NULL /*!80023 INVISIBLE */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0,008 sec)
```

Without the current patch, the invisible keyword will be discarded.

File: src/dialect/mod.rs [hunk 3]
```diff
@@ -898,6 +898,12 @@ pub trait Dialect: Debug + Any {
...
+    /// e.g. `/*! hint */`
+    fn supports_c_style_comments(&self) -> bool {
```
Comment (iffyio) [inline src/dialect/mod.rs]:
```suggestion
    fn supports_c_style_comment_hints(&self) -> bool {
```
Since c_style comment I imagine is referring to `/**/` which other dialects support?

File: src/tokenizer.rs (hunk 2 above)
Comment (iffyio) [reply src/tokenizer.rs:4280]: Ah so to clarify I'm rather wondering regarding the parser behavior for hints that aren't singe words e.g. `/*!50110 KEY_BLOCK_SIZE=1024 */` - can we demonstrate the behavior with test cases for such scenarios?

Comment (altmannmarcelo) [reply src/tokenizer.rs:4280]: @iffyio thanks flagging this. I have fixed the issue and now we properly return individual tokens inside a C-style hint comment.

File: src/dialect/mod.rs [hunk 4]
```diff
@@ -898,6 +898,12 @@ pub trait Dialect: Debug + Any {
         false
     }
 
+    /// Returns true if the dialect supports hint and C-style comments
+    /// e.g. `/*! hint */`
+    fn supports_c_style_hints(&self) -> bool {
```
Comment (iffyio) [inline src/dialect/mod.rs]:
```suggestion
    /// Returns true if the dialect supports optimizer hints in multiline comments
    /// e.g. `/*!50110 KEY_BLOCK_SIZE = 1024*/`
    fn supports_multiline_comment_hints(&self) -> bool {
```

File: src/tokenizer.rs [hunk 5]
```diff
@@ -936,10 +939,16 @@ impl<'a> Tokenizer<'a> {
 
     /// Get the next token or return None
     fn next_token(
-        &self,
+        &mut self,
         chars: &mut State,
         prev_token: Option<&Token>,
     ) -> Result<Option<Token>, TokenizerError> {
+        // Return any previously injected tokens first
```
Comment (iffyio) [inline src/tokenizer.rs]: impl wise I'm thinking we can do something like the following that would be less invasive to the tokenizer:

If we take as an example `/*!50110 KEY_BLOCK_SIZE = 1024*/` - today that will get parse as a `Token::MultilineComment("!50110 KEY_BLOCK_SIZE = 1024")` - which we receive [here in the tokenizer loop](https://github.com/altmannmarcelo/datafusion-sqlparser-rs/blob/f96249cc1d52d34edcfb0ca246bdd78f62552b91/src/tokenizer.rs#L907).

Then the idea would be to essentially re-tokenize that string and add to the tokens buffer.

So something like this:

```rust
// we check if this is a token containing optimizer hints
match token {
	Token::Whitespace(Whitespace::MultiLineComment(comment)) if self.dialect.supports_multiline_comment_hints() && s.starts_with("!") => {
    // re-tokenize the hints
	buf.extend(tokenize_comment_hints(comment, span)?);
	token => {
		buf.push(TokenWithSpan{ token, span })
	}
}

// here we reuse existing tokenizer machinery to re-tokenize the hints.
fn tokenize_comment_hints(&self, hints: String, span: Span, tokens: &mut Vec<TokenWithSpan>) -> Result<()> {
	let mut state = State {
		peekable: hints.chars().peekable(),
		line: span.start.line,
		col: span.start.column,
	};

	while let Some(token) = self.next_token(&mut state, None)? {
		tokens.push(token);
	}

	Ok(())
}
```

File: src/dialect/mod.rs (hunk 4 above)
Comment (iffyio) [reply src/dialect/mod.rs]: Ah to clarify the suggested change, c-style comment I think is rather understood as  `/* */` - i.e. multiline comment in general, the feature here I think is the optimizer hints within comments as mysql calls them

**Comment by altmannmarcelo** [comment]
I will pick this up in a few days.

**Comment by altmannmarcelo** [comment]
@iffyio sorry for my delay on getting back - I have implemented the re-tokenize function.

File: src/tokenizer.rs [hunk 6]
```diff
@@ -980,7 +1036,7 @@ impl<'a> Tokenizer<'a> {
 
     /// Get the next token or return None
     fn next_token(
-        &self,
+        &mut self,
```
Comment (iffyio) [inline src/tokenizer.rs]:
```suggestion
        &self,
```
the `mut` ref doesn't seem like it's being used?

File: src/tokenizer.rs [hunk 7]
```diff
@@ -2227,13 +2283,12 @@ impl<'a> Tokenizer<'a> {
     }
 
     fn tokenize_multiline_comment(
-        &self,
+        &mut self,
```
Comment (iffyio) [inline src/tokenizer.rs]:
```suggestion
        &self,
```
similarly, looks unused?

File: src/tokenizer.rs [hunk 8]
```diff
@@ -4218,6 +4273,47 @@ mod tests {
...
+                keyword: Keyword::NoKeyword,
+            }),
+            Token::Number("1".to_string(), false),
+        ];
+        compare(expected, tokens);
+    }
+
+    #[test]
+    fn tokenize_multiline_comment_with_comment_hint_and_version() {
+        let sql_multi = String::from("0 /*!50110 KEY_BLOCK_SIZE = 1024*/ 1");
```
Comment (iffyio) [inline src/tokenizer.rs:4300]: can we add a similar test cases where the comment has an empty hint? e.g.
```
0 /*!50110 */
0 /*!*/
0 /*!   */
```
